File: esp32s3_platformio/PYL Game S3/src/tuneplayer.cpp
#include "tuneplayer.h"

TunePlayer::TunePlayer(Buzzer& buzzer)
    : buzzer_(buzzer), cancelFlag_(false), tuneTaskHandle_(nullptr), isPlaying_(false) {}

TunePlayer::~TunePlayer() {
    stop();
}

void TunePlayer::play(const Tune& tune) {
    stop();  // Stop any existing playback
    cancelFlag_.store(false);
    isPlaying_.store(true);
    tune_ = tune;  // Copy the tune to the class member

    // Create a FreeRTOS task
    xTaskCreate(
        playTuneTask,       // Task function
        "TunePlayerTask",   // Name of the task
        4096,               // Stack size in words (adjust as needed)
        this,               // Task input parameter
        tskIDLE_PRIORITY,   // Priority (adjust as needed)
        &tuneTaskHandle_    // Task handle
    );
}

void TunePlayer::stop() {
    if (isPlaying()) {
        cancelFlag_.store(true);
        if (tuneTaskHandle_ != nullptr) {
            // Wait for the task to finish
            vTaskDelete(tuneTaskHandle_);
            tuneTaskHandle_ = nullptr;
        }
        isPlaying_.store(false);
    }
    buzzer_.stop();
}

bool TunePlayer::isPlaying() const {
    return isPlaying_.load();
}

void TunePlayer::playTuneTask(void* pvParameters) {
    TunePlayer* player = static_cast<TunePlayer*>(pvParameters);
    player->playTune();
    vTaskDelete(nullptr);  // Delete this task when done
}

void TunePlayer::playTuneShared() {
    int j = 0;
    int first = true;
    float multiplier = tune_.multiplier != 0 ? tune_.multiplier : 1.0f;
    for (int i = 0; i < tune_.notes.size(); ++i) {
        if (cancelFlag_.load()) {
            break;
        }

        int freq = tune_.notes[i].frequency;
        int duration = tune_.notes[i].duration;

        if (freq == 0 || freq == 8) {
            buzzer_.stop();
        } else {
            buzzer_.init(freq, 255);
        }

        if (tune_.debug) {
            printf("Note %d: Freq=%d, Duration=%dms\n", i, freq, duration);
        }

        // Use vTaskDelay instead of sleep
        sleepWithCancellation(static_cast<int>(duration / multiplier));
        if (tune_.callback) {
            if (tune_.callback_at_start && first && tune_.callback_interval != 1) {
                // don't double callback if interval is 1
                tune_.callback();
                first = false;
            } else {
                j++; // bump to 1 before interval comparison
                if (j == tune_.callback_interval) {
                    tune_.callback();
                    j = 0;
                }
            }
        }
    }
}

void TunePlayer::playTune() {
    try {
        if (tune_.loopable) {
            while (!cancelFlag_.load()) {
                playTuneShared();
            }
        } else {
            playTuneShared();
            buzzer_.stop();
        }
    } catch (...) {
        buzzer_.stop();
        isPlaying_.store(false);
        throw;
    }
    buzzer_.stop();
    isPlaying_.store(false);
}

void TunePlayer::sleepWithCancellation(int duration_ms) {
    const int interval = 50;
    int elapsed = 0;
    while (elapsed < duration_ms) {
        if (cancelFlag_.load()) {
            break;
        }
        int sleep_time = std::min(interval, duration_ms - elapsed);
        vTaskDelay(pdMS_TO_TICKS(sleep_time));
        elapsed += sleep_time;
    }
}

File: micropython/sounds/tuneplayer.py
import time, asyncio

# non-async for listening
def playtune(buzzer, tune):
    i = 0
    if 'multiplier' in tune:
        multiplier = tune['multiplier']
    else:
        multiplier = 1
        
    if tune['loopable']:
        while True:
            for i in range(0, len(tune['freqs'])):
                if tune['freqs'][i][0] == 0 or tune['freqs'][i][0] == 8:
                    buzzer.init(duty_u16=0)
                else:
                    buzzer.init(freq=tune['freqs'][i][0], duty_u16=32768)
                    
                timing = tune['freqs'][i][1]
                if 'debug' in tune:
                    if tune['debug'] == True:
                        print(i, tune['freqs'][i][0], timing)
                        i = i + 1

                time.sleep_ms(int(timing / multiplier))
    else:
        for i in range(0, len(tune['freqs'])):
            if tune['freqs'][i][0] == 0 or tune['freqs'][i][0] == 8:
                buzzer.init(duty_u16=0)
            else:
                buzzer.init(freq=tune['freqs'][i][0], duty_u16=32768)
                
            timing = tune['freqs'][i][1]

            if 'debug' in tune:
                if tune['debug'] == True:
                    print(i, tune['freqs'][i][0], timing)
                    i = i + 1

            time.sleep_ms(int(timing / multiplier))
        buzzer.init(duty_u16=0)

# async for game
async def playtune_async(buzzer, tune):
    i = 0
    if 'multiplier' in tune:
        multiplier = tune['multiplier']
    else:
        multiplier = 1
    try:
        if tune['loopable']:
            while True:
                for i in range(0, len(tune['freqs'])):
                    if tune['freqs'][i][0] == 0 or tune['freqs'][i][0] == 8:
                        buzzer.init(duty_u16=0)
                    else:
                        buzzer.init(freq=tune['freqs'][i][0], duty_u16=32768)
                        
                    timing = tune['freqs'][i][1]

                    if 'debug' in tune:
                        if tune['debug'] == True:
                            print(i, tune['freqs'][i][0], timing)
                            i = i + 1
                        
                    await asyncio.sleep_ms(int(timing / multiplier))
        else:
            for i in range(0, len(tune['freqs'])):
                if tune['freqs'][i][0] == 0 or tune['freqs'][i][0] == 8:
                    buzzer.init(duty_u16=0)
                else:
                    buzzer.init(freq=tune['freqs'][i][0], duty_u16=32768)
                    
                timing = tune['freqs'][i][1]

                if 'debug' in tune:
                    if tune['debug'] == True:
                        print(i, tune['freqs'][i][0], timing)
                        i = i + 1

                await asyncio.sleep_ms(int(timing / multiplier))
        buzzer.init(duty_u16=0)
    finally:
        buzzer.init(duty_u16=0)
